tests/ui/direct_recursion.rs

```diff
@@ -0,0 +1,112 @@
+#![deny(clippy::direct_recursion)]
+#![deny(clippy::allow_attributes_without_reason)]
+
+// Basic Cases //
+
+#[allow(unconditional_recursion, reason = "We're not testing for that lint")]
+fn i_call_myself_always() {
+    i_call_myself_always();
+    //~^ direct_recursion
+}
+
+fn i_call_myself_conditionally(do_i: bool) {
+    if do_i {
+        i_call_myself_conditionally(false);
+        //~^ direct_recursion
+    }
+}
+
+// Basic Counterexamples //
+
+fn i_get_called_by_others() {}
+
+fn i_call_something_else() {
+    i_get_called_by_others();
+}
+
+// Elaborate Cases //
+
+// Here we check that we're allowed to bless specific recursive calls.
+// A fine-grained control of where to allow recursion is desirable.
+// This is a test of such a feature.
+fn i_call_myself_in_a_bounded_way(bound: u8) {
+    if bound > 0 {
+        #[clippy::allowed_recursion]
```

Comment:
I don't understand the need for a dedicated attribute. If you are allowed to use it on an expression, you can allow the lint. If you want to allow it at a function level, then every inner recursive call, by construction, will inherit this allow, and won't be linted.